hudi-client/src/test/java/org/apache/hudi/TestHoodieClientOnCopyOnWriteStorage.java

```diff
@@ -382,6 +383,50 @@ public void testDeletes() throws Exception {
         HoodieWriteClient::upsert, true, 50, 150, 2);
   }
 
+  /**
+   * Test update of a record to different partition with Global Index.
+   */
+  @Test
+  public void testUpsertToDiffPartitionGlobalIndex() throws Exception {
+    HoodieWriteClient client = getHoodieWriteClient(getConfig(IndexType.GLOBAL_BLOOM), false);
+    /**
+     * Write 1 (inserts and deletes) Write actual 200 insert records and ignore 100 delete records
+     */
+    String initCommitTime = "000";
+    String newCommitTime = "001";
+    List<HoodieRecord> inserts1 = dataGen.generateInserts(newCommitTime, 10);
+
+    // Write 1 (only inserts)
+    client.startCommitWithTime(newCommitTime);
+    JavaRDD<HoodieRecord> writeRecords = jsc.parallelize(inserts1, 1);
+
+    JavaRDD<WriteStatus> result = client.insert(writeRecords, newCommitTime);
+    List<WriteStatus> statuses = result.collect();
+    assertNoWriteErrors(statuses);
+
+    // check the partition metadata is written out
+    assertPartitionMetadata(HoodieTestDataGenerator.DEFAULT_PARTITION_PATHS, fs);
+
+    /**
+     * Write 2 updates
+     */
+    String prevCommitTime = newCommitTime;
+    newCommitTime = "004";
+
+    // Write 1 (only inserts)
+    client.startCommitWithTime(newCommitTime);
+
+    List<HoodieRecord> updates1 = dataGen.generateUpdatesWithDiffPartition(newCommitTime, inserts1);
+    JavaRDD<HoodieRecord> updateRecords = jsc.parallelize(updates1, 1);
+
+    JavaRDD<WriteStatus> result1 = client.upsert(updateRecords, newCommitTime);
+    List<WriteStatus> statuses1 = result1.collect();
+    assertNoWriteErrors(statuses1);
+
+    // check the partition metadata is written out
+    assertPartitionMetadata(HoodieTestDataGenerator.DEFAULT_PARTITION_PATHS, fs);
```

Comment:
Does it enough to just check `PartitionMetadata`, how about comparing records?

Reply:
Most tests in this class does not verify records per se, just total records at each step. I will fix and do the same here.